### PR TITLE
Unrestricted boot

### PR DESCRIPTION
**Pull request by jreidinger**
Do not merge yet as grub2 is not yet ready. changelog update will be added later to avoid conflicts.

### CODE REVIEW

**Comment by mvidner** [comment]
> changelog update will be added later to avoid conflicts

Then you should state the intended changelog in the PR description, to get it reviewed. And I wouldn't have to ask for fate/bugzilla refrerence now ;-)

**Comment by jreidinger** [comment]
changelog is something like
```
- allow to select grub1 password protection behavior (fate#318574)
```

**Comment by mvidner** [comment]
> to select grub1 password protection
which version? "GRUB 2 " ?

File: src/lib/bootloader/grub2base.rb [hunk 1]
```diff
@@ -6,6 +6,9 @@
 module Yast
   # Common base for GRUB2 specialized classes
   class GRUB2Base < Module
+    # stored password object
```
Comment (mvidner) [inline src/lib/bootloader/grub2base.rb]: \# @return [::Bootloader::GRUB2Pwd] stored password object

File: src/lib/bootloader/grub2pwd.rb [hunk 2]
```diff
@@ -1,63 +1,131 @@
...
-  YAST_BASH_PATH = Yast::Path.new(".target.bash_output")
-  PWD_ENCRYPTION_FILE = "/etc/grub.d/42_password"
+Yast.import "Stage"
 
-  def used?
-    Yast.import "FileUtils"
+module Bootloader
+  # class is responsible for detection, encryption and writing of grub2 password protection
+  class GRUB2Pwd
+    attr_accessor :used, :unrestricted
```
Comment (mvidner) [inline src/lib/bootloader/grub2pwd.rb]: Please document that they are Boolean.
Password being "used" is IMHO obvious but "unrestricted" is not, and this is a good place to document what we mean by it.

File: src/lib/bootloader/grub2pwd.rb [hunk 3]
```diff
@@ -1,63 +1,131 @@
...
+    alias_method :used?, :used
+    alias_method :unrestricted?, :unrestricted
 
-    Yast::FileUtils.Exists PWD_ENCRYPTION_FILE
-  end
+    def initialize
+      # TODO: offline upgrade should somehow read it when we need grub2 -> grub2 modifications
```
Comment (mvidner) [inline src/lib/bootloader/grub2pwd.rb]: grub1 -> grub2 ?

File: src/lib/bootloader/grub2pwd.rb [hunk 4]
```diff
@@ -1,63 +1,131 @@
...
+      if used?
+        enable
+      else
+        disable
+      end
+    end
 
-  def disable
-    return unless used?
+    def password=(value)
```
Comment (mvidner) [inline src/lib/bootloader/grub2pwd.rb]: So, I can set a password, but not query its value, only its existence. OK, fair enough. But this is separated from the `used` switch? Shouldn't they be linked?

File: src/lib/bootloader/grub2pwd.rb [hunk 5]
```diff
@@ -1,63 +1,131 @@
...
 
-    if result["exit"] != 0
-      raise "Failed to create encrypted password for grub2. Command output: #{result["stderr"]}"
+      @encrypted_password = pwd_line[/password_pbkdf2 root (\S+)/, 1]
     end
 
-    pwd_line = result["stdout"].split("\n").grep(/password is/).first
-    if !pwd_line
-      raise "INTERNAL ERROR: output do not contain encrypted password. Output: #{result["stdout"]}"
+    def read_used
```
Comment (mvidner) [inline src/lib/bootloader/grub2pwd.rb]: rename to `read_used?` to make clear it's a Boolean

File: src/lib/bootloader/grub2pwd.rb (hunk 3 above)
Comment (jreidinger) [inline src/lib/bootloader/grub2pwd.rb]: for grub1 -> grub2 we do repropose, password config is lost and cannot be converted as it use different hashing algorithm

File: src/lib/bootloader/grub2pwd.rb (hunk 4 above)
Comment (jreidinger) [inline src/lib/bootloader/grub2pwd.rb]: reason is that enabling password do not necessary mean, that you specify password. So used != password set...also it mean when in UI you set used = false and then go back and set used = true, it remember old password which is set to widgets.

File: src/lib/bootloader/grub2pwd.rb (hunk 2 above)
Comment (jreidinger) [inline src/lib/bootloader/grub2pwd.rb]: yep, I forgot to write api documentation, my fault.

**Comment by mvidner** [comment]
- Let password protection be configurable between a restricted
  mode (cannot boot at all without password, default GRUB2
  behavior) and an unrestricted mode (can boot but cannot select
  or edit entries, GRUB1 behavior) (FATE#318574).

File: src/lib/bootloader/grub2pwd.rb [hunk 6]
```diff
@@ -1,63 +1,131 @@
...
+      Yast::FileUtils.Exists PWD_ENCRYPTION_FILE
     end
 
-    ret = pwd_line[/^.*password is\s*(\S+)/, 1]
-    if !ret
-      raise "INTERNAL ERROR: output do not contain encrypted password. Output: #{result["stdout"]}"
+    def enable
+      raise "Wrong code: password not written" unless @encrypted_password
+
+      file_content = "#! /bin/sh\n" \
```
Comment (mvidner) [inline src/lib/bootloader/grub2pwd.rb]: > \# The files in /etc/grub.d are programs that write GRUB 2 programs on their stdout.
\# So 'exec tail' is a way of saying "just echo the rest of this program as its output".

Comment (mvidner) [inline src/lib/bootloader/grub2pwd.rb]: I mean this as a comment in the rb file, not in the config file (or its output!!)

File: src/lib/bootloader/grub2pwd.rb (hunk 5 above)
Comment (mvidner) [inline src/lib/bootloader/grub2pwd.rb]: So the intent is "read the value for the attribute `used` from permanent storage but don't assign it". 

The yast convention is that Read should assign the attributes (and the preceding methods conforms to that).

Maybe we can set a convention that Read without assigning is called Load. -> `load_used`

File: src/lib/bootloader/grub2pwd.rb (hunk 4 above)
Comment (mvidner) [inline src/lib/bootloader/grub2pwd.rb]: So the state "password is set but disabled" can exist, but only in the UI, not in the config files, right?

File: src/lib/bootloader/grub2pwd.rb [hunk 7]
```diff
@@ -1,63 +1,146 @@
...
-  def used?
-    Yast.import "FileUtils"
+module Bootloader
+  # class is responsible for detection, encryption and writing of grub2 password protection
+  class GRUB2Pwd
+    # @!attribute used
+    #   @return [bool] specifies if password protection enabled
+    #
+    # @!attribute unrestricted
+    #   @return [bool] specifies if unrestricted password protection should be used
```
Comment (mvidner) [inline src/lib/bootloader/grub2pwd.rb]: "Boolean", not "bool".
Again, somewhere we should say what "unrestricted" means and I think it should be here.

Comment (jreidinger) [inline src/lib/bootloader/grub2pwd.rb]: even Boolean is not known class in ruby, there is only TrueClass and FalseClass.
Well, unrestricted is official grub2 term. I expect that this code will be modified only by someone who knows grub2 terminology.

File: src/include/bootloader/grub/options.rb [hunk 8]
```diff
@@ -71,14 +72,22 @@ def passwd_content
...
-            Password(Id(:pw1), Opt(:hstretch), _("&Password")),
-            # text entry
-            HSpacing(2),
-            Password(Id(:pw2), Opt(:hstretch), _("Re&type Password")),
-            HStretch()
+          VBox(
+            HBox(
+              HSpacing(2),
+              # TRANSLATORS: checkbox entry
+              CheckBox(Id(:unrestricted_pw), _("A&llow to boot locked default entry without password")),
```
Comment (mvidner) [inline src/include/bootloader/grub/options.rb]: Why are we talking about a "locked" default entry? I thought being locked is a global property for all entries together, controlled by the "Password" checkbox.

File: test/grub2pwd_test.rb [hunk 9]
```diff
@@ -55,52 +221,47 @@ def mock_file_presence(exists)
...
-    it "raise exception if grub2-mkpasswd-pbkdf create password line but without password" do
-      expect(Yast::WFM).to receive(:Execute)
-        .with(kind_of(Yast::Path), /grub2-mkpasswd/)
-        .and_return(
-          "exit"   => 0,
-          "stderr" => "",
-          "stdout" => "password is"
-        )
-      expect(Yast::SCR).to receive(:Write).never
+    it "returns true if password explicitelly set" do
```
Comment (mvidner) [inline test/grub2pwd_test.rb]: explicitly

File: src/lib/bootloader/grub2pwd.rb (hunk 7 above)
Comment (mvidner) [inline src/lib/bootloader/grub2pwd.rb]: Yes, but Boolean is a YARD convention whereas bool is not: http://www.rubydoc.info/gems/yard/file/docs/GettingStarted.md#Declaring_Types

Yes, "unrestricted" is an official grub2 term, but it is local to a single entry whereas we use it in a global meaning. Even if I read the docs I will not know what it means is SUSE: https://www.gnu.org/software/grub/manual/html_node/Security.html#Security

File: src/lib/bootloader/grub2pwd.rb [hunk 10]
```diff
@@ -1,63 +1,146 @@
...
+    attr_accessor :used, :unrestricted
+    alias_method :used?, :used
+    alias_method :unrestricted?, :unrestricted
 
-    Yast::FileUtils.Exists PWD_ENCRYPTION_FILE
-  end
+    # Reads or proposes configuration depending on stage
+    def initialize
+      # TODO: offline upgrade should somehow read it when we need grub2 -> grub2 modifications
```
Comment (mvidner) [inline src/lib/bootloader/grub2pwd.rb]: IMHO this comment is anticipating future requirements, YAGNI, delete it.

File: src/lib/bootloader/grub2pwd.rb [hunk 11]
```diff
@@ -1,63 +1,146 @@
...
+    end
+
+  private
+
+    YAST_BASH_PATH = Yast::Path.new(".target.bash_output")
+    PWD_ENCRYPTION_FILE = "/etc/grub.d/42_password"
+
+    def propose
+      @used = false
+      @unrestricted = false # TODO: ensure it in FATE
```
Comment (mvidner) [inline src/lib/bootloader/grub2pwd.rb]: ... whereas this TODO should be resolved before merging. (and then deleted).

File: src/include/bootloader/grub/options.rb (hunk 8 above)
Comment (jreidinger) [inline src/include/bootloader/grub/options.rb]: I want to emphasis that it is only default entry without possibility to modify kernel parameters

File: src/lib/bootloader/grub2pwd.rb (hunk 11 above)
Comment (jreidinger) [inline src/lib/bootloader/grub2pwd.rb]: fate is already in needinfo state so I need to wait

File: src/include/bootloader/grub/helps.rb [hunk 12]
```diff
@@ -80,8 +80,12 @@ def initialize_bootloader_grub_helps(_include_target)
...
         "password"          => _(
           "<p><b>Protect Boot Loader with Password</b><br>\n" \
-            "Define the password that will be required to access the boot menu. YaST will only accept the password if you repeat\n" \
-            "it in <b>Retype Password</b>.</p>"
+            "Define the password that will be required to access the boot menu. \n" \
+            "YaST will only accept the password if you repeat\n" \
+            "it in <b>Retype Password</b>. If <b>Allow to boot locked default \n" \
+            "entry without password</b> is checked, then without password is \n " \
+            "possible to let bootloader autoboot default entry with its kernel \n" \
+            "parameters.</p>"
```
Comment (mvidner) [inline src/include/bootloader/grub/helps.rb]: > **Protect Boot Loader with Password**
At boot time, modifying or even booting any entry will require the password. If **Allow to Boot Default Entry** is checked then that action will not require the password.
YaST will only accept the password if you repeat it in **Retype Password.**

(is Entry the correct term? Section?)

Comment (arvidjaar) [inline src/include/bootloader/grub/helps.rb]: > If Allow to Boot Default Entry is checked then that action will not require the password.
May be it helps to understand what GRUB2 actually implements.

You can enable authentication in GRUB2 (by setting superusers variable to list of user names). As soon as superusers variable is defined, entering CLI, editing menu entry and executing menu entry (i.e. usually booting) requires authentication as one of users in this list.

It is possible to either provide additional list of users that are allowed to execute individual menu entry (but not to edit it) or mark individual menu entry as "unrestricted" so executing it does not require authentication.

What currently is *NOT* possible is to autoboot default entry while keeping password protection for it. Either entry must be marked as unrestricted (and so always allowed to boot, manually or auto) or you need to enter password both for manual or autoboot.

Looking at it, autoboot without password looks like sensible feature request.

Comment (jreidinger) [inline src/include/bootloader/grub/helps.rb]: Well, request from our customer is to option to have back grub1 pwd protection, I need to verify how exactly it works in grub1 and compare it to testing grub2 package that michael provide to me.

Comment (jreidinger) [inline src/include/bootloader/grub/helps.rb]: I check grub1 feature and it also allow to boot any boot entry, only lock kernel parameters

Comment (jreidinger) [inline src/include/bootloader/grub/helps.rb]: and for grub2 it behave same way. @mvidner can you look at it and propose possibly new labels and help text for it, that make better sense? something like
```
label = "Protect entry modification only"
helptext = "<b> Protect entry modification only</b> protects by password only modification of entry like loading different kernel or modify kernel parameters."
```

Comment (mvidner) [inline src/include/bootloader/grub/helps.rb]: @arvidjaar Thanks for the clarification! We are implementing a GRUB1 compatibility switch and I originally thought it did what I described above. Here's a revised description:

> **Protect Boot Loader with Password**
At boot time, modifying or even booting any entry will require the password. If **Protect Entry Modification Only** is checked then booting any entry is not restricted but modifying entries requires the password.
YaST will only accept the password if you repeat it in **Retype Password.**

File: src/lib/bootloader/grub2pwd.rb (hunk 11 above)
Comment (mvidner) [inline src/lib/bootloader/grub2pwd.rb]: Still pending.

**Comment by mvidner** [comment]
Code looks good but the changelog entry is missing:
> - Let password protection be configurable between a restricted
  mode (cannot boot at all without password, default GRUB2
  behavior) and an unrestricted mode (can boot but cannot edit entries,
  GRUB1 behavior) (FATE#318574).

**Comment by jreidinger** [comment]
for changelog I still waiting for michael to submit new grub2 to SP1, when he add it I also I add it.

File: src/include/bootloader/grub/helps.rb [hunk 13]
```diff
@@ -80,8 +80,12 @@ def initialize_bootloader_grub_helps(_include_target)
         ),
         "password"          => _(
           "<p><b>Protect Boot Loader with Password</b><br>\n" \
-            "Define the password that will be required to access the boot menu. YaST will only accept the password if you repeat\n" \
-            "it in <b>Retype Password</b>.</p>"
+            "At boot time, modifying or even booting any entry will require the" \
+            " password. If <b>Protect Entry Modification Only</b> is checked then " \
+            "booting any entry is not restricted but modifying entries requires " \
+            "the password.<br>" \
```
Comment (mvidner) [inline src/include/bootloader/grub/helps.rb]: "the password (which is the way GRUB 1 behaved).\<br>"

File: src/include/bootloader/grub/helps.rb [hunk 14]
```diff
@@ -80,8 +80,12 @@ def initialize_bootloader_grub_helps(_include_target)
...
+            "YaST will only accept the password if you repeat it in " \
+            "<b>Retype Password</b>."
```
Comment (mvidner) [inline src/include/bootloader/grub/helps.rb]: \</p>

**Comment by mvidner** [comment]
LGTM!

**Pull request merged**